src/main/java/org/folio/event/handler/HoldingUpdateAsyncRecordHandler.java

```diff
@@ -0,0 +1,193 @@
+      log.warn("updateAdjacentHoldingsWithNewInstanceId:: No adjacent holdings were found to update, ignoring update");
+      return Future.succeededFuture();
+    }
+    var poLineFutures = new ArrayList<Future<Void>>();
+    var requestHeaders = new HeadersMultiMap().addAll(headers);
+    var okapiUrl = headers.get(XOkapiHeaders.URL);
+    var newInstanceId = JsonObject.mapFrom(resourceEvent.getNewValue()).getString(INSTANCE_ID);
+    holdingIds.forEach(holdingId -> poLineFutures.add(updateAdjacentHolding(okapiUrl, requestHeaders, holdingId, newInstanceId)));
+    return GenericCompositeFuture.all(poLineFutures)
+      .onComplete(asyncResult -> log.info("updateAdjacentHoldingsWithNewInstanceId:: Updated adjacent holdings, size: {}", holdingIds.size()))
+      .mapEmpty();
+  }
+
+  private Future<Void> updateAdjacentHolding(String okapiUrl, MultiMap requestHeaders, String holdingId, String newInstanceId) {
+    log.info("updateAdjacentHolding:: Updating an adjacent holding, holdingId: {}, instanceId: {}", holdingId, newInstanceId);
+    var absoluteUri = String.format(STORAGE_HOLDING_URL, okapiUrl, holdingId);
+    return webClient.requestAbs(HttpMethod.GET, absoluteUri)
+      .putHeaders(requestHeaders)
+      .send()
+      .compose(getResult ->
+        webClient.requestAbs(HttpMethod.PUT, absoluteUri)
+          .putHeaders(requestHeaders)
+          .sendJson(getResult.bodyAsJsonObject().put(INSTANCE_ID, newInstanceId))
+          .onSuccess(putResult -> log.info("updateAdjacentHolding:: Updated an adjacent holding, holdingId: {}, statusCode: {}",
+            holdingId, putResult.statusCode()))
+          .onFailure(throwable -> log.error("updateAdjacentHolding:: Failed to update an adjacent holding, holdingId: {}", holdingId, throwable))
+          .mapEmpty());
+  }
+
+  private boolean updatePoLinesInstanceId(ResourceEvent resourceEvent, List<PoLine> poLines) {
+    var pair = getIdPair(resourceEvent, INSTANCE_ID);
+    if (pair.getLeft().equals(pair.getRight())) {
+      log.warn("updatePoLinesInstanceId:: No instance id was changed (ids are the same), ignoring update");
+      return false;
+    }
+    for (var poline : poLines) {
+      poline.setInstanceId(pair.getRight());
+      log.info("updatePoLinesInstanceId:: Added new instance id to POL, poLineId: {}, old instanceId: {}, new instanceId: {}",
+        poline.getId(), pair.getLeft(), pair.getRight());
+    }
+    return true;
+  }
+
+  private boolean updatePoLinesSearchLocationIds(ResourceEvent resourceEvent, List<PoLine> poLines) {
+    var pair = getIdPair(resourceEvent, PERMANENT_LOCATION_ID);
+    if (pair.getLeft().equals(pair.getRight())) {
+      log.warn("updatePoLinesSearchLocationIds:: No search location id was changed (ids are the same), ignoring update");
+      return false;
+    }
+    for (var poline : poLines) {
+      var searchLocationIds = poline.getSearchLocationIds();
+      if (searchLocationIds.remove(pair.getLeft())) {
```

Comment:
need to think can we remove searchLocationId, because there can be case when this searchLocationId is used also by other holding connected to the same po line

Reply:
Removed id deletion as planned